File: rust/src/stream.rs
mod establish_connection;
mod monitor_connection;

use establish_connection::connect;
use monitor_connection::run_stream;

use crate::{config::Config, feed::ID, report::Report};

use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    sync::{
        atomic::{AtomicU64, Ordering},
        Arc,
    },
};
use tokio::{
    net::TcpStream,
    sync::{broadcast, mpsc, Mutex},
    time::{sleep, Duration},
};
use tokio_tungstenite::{MaybeTlsStream, WebSocketStream as TungsteniteWebSocketStream};
use tracing::{error, info};

#[derive(Debug, thiserror::Error)]
pub enum StreamError {
    #[error("WebSocket error: {0}")]
    WebSocketError(#[from] tokio_tungstenite::tungstenite::Error),

    #[error("Connection error: {0}")]
    ConnectionError(String),

    #[error("Authentication error: {0}")]
    AuthError(#[from] crate::auth::HmacError),

    #[error("Serialization error: {0}")]
    SerializationError(#[from] serde_json::Error),

    #[error("Stream closed")]
    StreamClosed,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WebSocketReport {
    pub report: Report,
}

struct Stats {
    /// Total number of accepted reports
    accepted: AtomicU64,
    /// Total number of deduplicated reports when in HA           
    deduplicated: AtomicU64,
    /// Total number of partial reconnects when in HA        
    partial_reconnects: AtomicU64,
    /// Total number of full reconnects    
    full_reconnects: AtomicU64,
    /// Number of configured connections if in HA      
    configured_connections: AtomicU64,
    /// Current number of active connections     
    active_connections: AtomicU64,
}

#[derive(Debug)]
pub enum WebSocketConnection {
    Single(TungsteniteWebSocketStream<MaybeTlsStream<TcpStream>>),
    Multiple(Vec<TungsteniteWebSocketStream<MaybeTlsStream<TcpStream>>>),
}

/// Stream represents a realtime report stream.
/// Safe for concurrent usage.
/// When HA mode is enabled and at least 2 origins are provided, the Stream will maintain at least 2 concurrent connections to different instances
/// to ensure high availability, fault tolerance and minimize the risk of report gaps.
pub struct Stream {
    config: Config,
    feed_ids: Vec<ID>,
    conn: Option<WebSocketConnection>,
    report_sender: mpsc::Sender<WebSocketReport>,
    report_receiver: mpsc::Receiver<WebSocketReport>,
    shutdown_sender: broadcast::Sender<()>,
    stats: Arc<Stats>,
    water_mark: Arc<Mutex<HashMap<String, usize>>>,
}

impl Stream {
    /// Establishes a streaming WebSocket connection that sends reports for the given feedID(s) after they are verified.
    ///
    /// # Arguments
    ///
    /// * `config` - A validated `Config` instance.
    /// * `feedIDs` - A comma-separated list of Data Streams feed IDs.
    ///
    /// # Endpoint:
    /// ```bash
    /// /api/v1/ws
    /// ```
    ///
    /// # Type:
    /// * WebSocket
    ///
    /// # Sample Request:
    /// ```bash
    /// GET /api/v1/ws?feedIDs=<feedID1>,<feedID2>,...
    /// ```
    ///
    /// # Sample Response:
    /// ```json
    /// {
    ///     "report": {
    ///         "feedID": "Hex encoded feedId.",
    ///         "fullReport": "A blob containing the report context and body. Encode the fee token into the payload before passing it to the contract for verification.",
    ///         "validFromTimestamp": "Report's earliest applicable timestamp (in seconds).",
    ///         "observationsTimestamp": "Report's latest applicable timestamp (in seconds)."
    ///     }
    /// }
    /// ```
    ///
    /// # Error Response Codes
    ///
    /// | Status Code | Description |
    /// |-------------|-------------|
    /// | **400 Bad Request** | This error is triggered when:<br>- There is any missing/malformed query argument.<br>- Required headers are missing or provided with incorrect values. |
    /// | **401 Unauthorized User** | This error is triggered when:<br>- Authentication fails, typically because the HMAC signature provided by the client doesn't match the one expected by the server.<br>- A user requests access to a feed without the appropriate permission or that does not exist. |
    /// | **500 Internal Server** | Indicates an unexpected condition encountered by the server, preventing it from fulfilling the request. This error typically points to issues on the server side. |
    pub async fn new(config: &Config, feed_ids: Vec<ID>) -> Result<Stream, StreamError> {
        let (report_sender, report_receiver) = mpsc::channel(100);
        let (shutdown_sender, _) = broadcast::channel(1);

        let stats = Arc::new(Stats {
            accepted: AtomicU64::new(0),
            deduplicated: AtomicU64::new(0),
            partial_reconnects: AtomicU64::new(0),
            full_reconnects: AtomicU64::new(0),
            configured_connections: AtomicU64::new(0),
            active_connections: AtomicU64::new(0),
        });

        let conn = connect(config, &feed_ids, stats.clone()).await?;

        let water_mark = Arc::new(Mutex::new(HashMap::new()));

        Ok(Stream {
            config: config.clone(),
            feed_ids,
            conn: Some(conn),
            report_sender,
            report_receiver,
            shutdown_sender,
            stats,
            water_mark,
        })
    }

    /// Starts listening for reports on the Stream.
    /// This method will spawn a new task for each WebSocket connection.
    pub async fn listen(&mut self) -> Result<(), StreamError> {
        let conn = self
            .conn
            .take()
            .ok_or_else(|| StreamError::ConnectionError("No connection".into()))?;

        match conn {
            WebSocketConnection::Single(stream) => {
                let report_sender = self.report_sender.clone();
                let shutdown_receiver = self.shutdown_sender.subscribe();
                let stats = self.stats.clone();
                let water_mark = self.water_mark.clone();
                let config = self.config.clone();
                let feed_ids = self.feed_ids.clone();

                tokio::spawn(run_stream(
                    stream,
                    report_sender,
                    shutdown_receiver,
                    stats,
                    water_mark,
                    config,
                    feed_ids,
                ));
            }
            WebSocketConnection::Multiple(streams) => {
                for stream in streams {
                    let report_sender = self.report_sender.clone();
                    let shutdown_receiver = self.shutdown_sender.subscribe();
                    let stats = self.stats.clone();
                    let water_mark = self.water_mark.clone();
                    let config = self.config.clone();
                    let feed_ids = self.feed_ids.clone();

                    tokio::spawn(run_stream(
                        stream,
                        report_sender,
                        shutdown_receiver,
                        stats,
                        water_mark,
                        config,
                        feed_ids,
                    ));
                }
            }
        }

        Ok(())
    }

    /// Reads the next available report on the Stream.
    /// Reads blocks until a report is received, the context is canceled or all underlying connections are in a error state.
    ///
    /// # Returns
    ///
    /// * `WebSocketReport` - The next available report.
    pub async fn read(&mut self) -> Result<WebSocketReport, StreamError> {
        self.report_receiver
            .recv()
            .await
            .ok_or(StreamError::StreamClosed)
    }

    /// Closes the Stream.
    /// It is the caller's responsibility to call close when the stream is no longer needed.
    pub async fn close(&mut self) -> Result<(), StreamError> {
        info!("Closing stream...");

        // Send shutdown signal
        if let Err(e) = self.shutdown_sender.send(()) {
            error!("Error sending shutdown signal: {:?}", e);
        }

        // Allow tasks to shut down gracefully
        sleep(Duration::from_millis(100)).await;

        Ok(())
    }

    /// Returns basic stats about the Stream.
    ///
    /// # Returns
    ///
    /// * `StatsSnapshot` - A snapshot of the current Stream statistics.
    ///     * `accepted` - Total number of accepted reports.
    ///     * `deduplicated` - Total number of deduplicated reports when in HA.
    ///     * `total_received` - Total number of received reports.
    ///     * `partial_reconnects` - Total number of partial reconnects when in HA.
    ///     * `full_reconnects` - Total number of full reconnects.
    ///     * `configured_connections` - Number of configured connections if in HA.
    ///     * `active_connections` - Current number of active connections.
    pub fn get_stats(&self) -> StatsSnapshot {
        let accepted = self.stats.accepted.load(Ordering::SeqCst);
        let deduplicated = self.stats.deduplicated.load(Ordering::SeqCst);

        StatsSnapshot {
            accepted,
            deduplicated,
            total_received: accepted + deduplicated,
            partial_reconnects: self.stats.partial_reconnects.load(Ordering::SeqCst),
            full_reconnects: self.stats.full_reconnects.load(Ordering::SeqCst),
            configured_connections: self.stats.configured_connections.load(Ordering::SeqCst),
            active_connections: self.stats.active_connections.load(Ordering::SeqCst),
        }
    }
}

/// Snapshot of statistics for external consumption.
#[derive(Debug, Clone)]
pub struct StatsSnapshot {
    /// Total number of accepted reports
    pub accepted: u64,
    /// Total number of deduplicated reports when in HA
    pub deduplicated: u64,
    /// Total number of received reports
    pub total_received: u64,
    /// Total number of partial reconnects when in HA
    pub partial_reconnects: u64,
    /// Total number of full reconnects
    pub full_reconnects: u64,
    /// Number of configured connections if in HA
    pub configured_connections: u64,
    /// Current number of active connections
    pub active_connections: u64,
}

File: go/stream.go
package streams

import (
	"context"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"github.com/smartcontractkit/data-streams-sdk/go/feed"
	"nhooyr.io/websocket"
)

const (
	defaultWSConnectTimeout      = time.Second * 5
	minWSReconnectIntervalMillis = 1000
	maxWSReconnectIntervalMIllis = 10000
	maxWSReconnectAttempts       = 5
)

var (
	ErrStreamClosed = fmt.Errorf("client: use of closed Stream")
)

type message struct {
	Report *ReportResponse `json:"report"`
}

// Stream represents a realtime report stream.
// Safe for concurrent usage.
//
// The Stream will maintain at least 2 concurrent connections to different instances
// to ensure high availability, fault tolerance and minimize the risk of report gaps.
type Stream interface {
	// Read the next available report on the Stream.
	// Read blocks until a report is received, the context is canceled or
	// all underlying connections are in a error state.
	Read(context.Context) (*ReportResponse, error)

	// Stats return basic stats about the Stream.
	Stats() Stats

	// Close the Stream. Is the caller responsibility to call close when
	// the stream is no longer needed.
	Close() error
}

// Stats for the Stream
type Stats struct {
	Accepted              uint64 // Total number of accepted reports
	Deduplicated          uint64 // Total number of deduplicated reports when in HA
	TotalReceived         uint64 // Total number of received reports
	PartialReconnects     uint64 // Total number of partial reconnects when in HA
	FullReconnects        uint64 // Total number of full reconnects
	ConfiguredConnections uint64 // Number of configured connections if in HA
	ActiveConnections     uint64 // Current number of active connections
}

func (s Stats) String() (st string) {
	return fmt.Sprintf(
		"accepted: %d, deduplicated: %d, total_received %d, partial_reconnects: %d, full_reconnects: %d, configured_connections: %d, active_connections %d",
		s.Accepted, s.Deduplicated,
		s.TotalReceived, s.PartialReconnects,
		s.FullReconnects, s.ConfiguredConnections, s.ActiveConnections,
	)
}

type stream struct {
	httpClient         *http.Client
	customHeaders      http.Header
	config             Config
	output             chan *ReportResponse
	feedIDs            []feed.ID
	conns              []*wsConn
	streamCtx          context.Context
	streamCtxCancel    context.CancelFunc
	closeError         atomic.Value
	connStatusCallback func(isConneccted bool, host string, origin string)

	waterMarkMu sync.Mutex
	waterMark   map[string]uint64

	stats struct {
		accepted              atomic.Uint64
		skipped               atomic.Uint64
		partialReconnects     atomic.Uint64
		fullReconnects        atomic.Uint64
		activeConnections     atomic.Uint64
		configuredConnections atomic.Uint64
	}

	closed       atomic.Bool
	closingMutex sync.RWMutex
}

func (c *client) newStream(ctx context.Context, httpClient *http.Client, feedIDs []feed.ID,
	origins []string, connStatusCallback func(isConnected bool, host string, origin string)) (s *stream, err error) {
	streamCtx, streamCtxCancel := context.WithCancel(ctx)
	s = &stream{
		httpClient:         httpClient,
		connStatusCallback: connStatusCallback,
		config:             c.config,
		output:             make(chan *ReportResponse, 1),
		feedIDs:            feedIDs,
		waterMark:          make(map[string]uint64),
		streamCtx:          streamCtx,
		streamCtxCancel:    streamCtxCancel,
	}

	if value := ctx.Value(CustomHeadersCtxKey); value != nil {
		if h, ok := value.(http.Header); ok {
			s.customHeaders = h
		}
	}

	// only creates a HA stream if
	// more than a single origin is provided
	// and ws ha is enabled
	if len(origins) > 0 && c.config.WsHA {
		c.config.logDebug("client: attempting to connect websockets in HA mode")
		var errs []error
		for x := 0; x < len(origins); x++ {
			s.stats.configuredConnections.Add(1)
			conn, err := s.newWSconn(ctx, origins[x])
			if err != nil {
				c.config.logInfo("client: failed to connect to origin %s: %s", origins[x], err)
				errs = append(errs, fmt.Errorf("origin %s: %w", origins[x], err))
				// Retry connecting to the origin in the background
				go func() {
					conn, err := s.newWSconnWithRetry(origins[x])
					if err != nil {
						return
					}
					go s.monitorConn(conn)
					// Lock is aquired here to prevent race condition with Close() occuring
					// during this background reconnect attempt
					s.closingMutex.Lock()
					s.conns = append(s.conns, conn)
					s.closingMutex.Unlock()
				}()
				continue
			}
			go s.monitorConn(conn)
			s.conns = append(s.conns, conn)
		}

		// Only fail if we couldn't connect to ANY origins
		if len(s.conns) == 0 {
			err = fmt.Errorf("failed to connect to any origins in HA mode: %v", errs)
			s.closeError.CompareAndSwap(nil, err)
			s.Close()
			return nil, err
		}
		c.config.logInfo("client: connected to %d out of %d origins in HA mode", len(s.conns), len(origins))
	} else {
		conn, err := s.newWSconn(ctx, "")
		if err != nil {
			return nil, err
		}
		go s.monitorConn(conn)
		s.conns = append(s.conns, conn)
		s.stats.configuredConnections.Add(1)
	}

	return s, nil
}

func (s *stream) pingConn(ctx context.Context, conn *wsConn) {
	ticker := time.NewTicker(time.Second * 2)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return

		case <-ticker.C:
			pctx, pcancel := context.WithTimeout(context.Background(), 2*time.Second)
			err := conn.conn.Ping(pctx)
			pcancel()

			if s.closed.Load() {
				return
			}

			if err != nil {
				s.config.logInfo(
					"client: stream websocket %s ping error: %s, closing client: %s",
					conn.origin, err, conn.close(),
				)
				return
			}
		}
	}
}

func (s *stream) monitorConn(conn *wsConn) {
	if s.connStatusCallback != nil {
		go s.connStatusCallback(true, conn.host, conn.origin)
	}
	for !s.closed.Load() {
		ctx, cancel := context.WithCancel(s.streamCtx)

		// start pinging the server in the background and ensure we fail
		// an unresponsive connection fast
		go s.pingConn(ctx, conn)

		// Set this conn to active
		s.stats.activeConnections.Add(1)

		// read blocks until conn is closed or errors out
		err := conn.read(ctx, &s.closingMutex, s.accept)
		cancel()
		// `Add(^uint64(0))` will decrement activeConnections
		s.stats.activeConnections.Add(^uint64(0))
		if s.connStatusCallback != nil {
			go s.connStatusCallback(false, conn.host, conn.origin)
		}

		// check for stream close conditions before reconnect attempts
		if ctxErr := s.streamCtx.Err(); ctxErr != nil || s.closed.Load() {
			if ctxErr != nil {
				s.config.logInfo(
					"client: stream websocket %s context done: %s",
					conn.origin, s.streamCtx.Err(),
				)
				conn.close()
			}
			return
		}

		// reconnect protocol
		if s.stats.activeConnections.Load() == 0 {
			s.stats.fullReconnects.Add(1)
		} else {
			s.stats.partialReconnects.Add(1)
		}

		s.config.logInfo(
			"client: stream websocket %s error: %s",
			conn.origin, err,
		)
		s.config.logInfo(
			"client: reconnecting stream websocket %s",
			conn.origin,
		)

		// ensure the current connection is closed
		_ = conn.close()

		re, err := s.newWSconnWithRetry(conn.origin)
		if err != nil {
			s.closeError.CompareAndSwap(nil, fmt.Errorf("stream has no active connections, last error: %w", err))
			s.Close()
			return
		}
		conn.replace(re.conn)
		if s.connStatusCallback != nil {
			go s.connStatusCallback(true, conn.host, conn.origin)
		}
		s.config.logInfo(
			"client: stream websocket %s: reconnected",
			conn.origin,
		)
	}
}

func (s *stream) newWSconnWithRetry(origin string) (conn *wsConn, err error) {
	// reconnect loop
	// will try to reconnect until client is closed or
	// we have no active connections and have exceeded maxWSReconnectAttempts
	var attempts int
	for {
		if s.closed.Load() || s.streamCtx.Err() != nil {
			return nil, fmt.Errorf("Retry cancelled, stream is closed")
		}

		// fail the stream if we are over the maxWSReconnectAttempts
		// and there are no other active connection
		if attempts >= s.config.WsMaxReconnect && s.stats.activeConnections.Load() == 0 {
			return nil, err
		}
		attempts++

		ctx, cancel := context.WithTimeout(context.Background(), defaultWSConnectTimeout)
		conn, err = s.newWSconn(ctx, origin)
		cancel()

		if err != nil {
			interval := time.Millisecond * time.Duration(
				rand.Intn(maxWSReconnectIntervalMIllis-minWSReconnectIntervalMillis)+minWSReconnectIntervalMillis) //nolint:gosec
			s.config.logInfo(
				"client: stream websocket %s: error reconnecting: %s, backing off: %s",
				origin, err, interval.String(),
			)
			time.Sleep(interval)
			continue
		}
		return conn, nil
	}
}

func (s *stream) Stats() (st Stats) {
	st.Accepted = s.stats.accepted.Load()
	st.Deduplicated = s.stats.skipped.Load()
	st.TotalReceived = st.Accepted + st.Deduplicated
	st.PartialReconnects = s.stats.partialReconnects.Load()
	st.FullReconnects = s.stats.fullReconnects.Load()
	st.ConfiguredConnections = s.stats.configuredConnections.Load()
	st.ActiveConnections = s.stats.activeConnections.Load()

	return st
}

func (s *stream) Read(ctx context.Context) (r *ReportResponse, err error) {
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case r := <-s.output:
		// return a valid report even if stream is closed
		// errstreamClosed will be returned on next call to Read()
		if s.closed.Load() && r == nil {
			if err, ok := s.closeError.Load().(error); ok {
				return nil, err
			}
			return nil, ErrStreamClosed
		}
		return r, nil
	}
}

func (s *stream) Close() (err error) {
	if !s.closed.CompareAndSwap(false, true) {
		return nil
	}
	s.streamCtxCancel()
	// this lock ensures websocket readers stop in a safe spot for closing
	s.closingMutex.Lock()
	defer s.closingMutex.Unlock()

	for x := 0; x < len(s.conns); x++ {
		_ = s.conns[x].close()
	}
	close(s.output)
	// return a pending error
	if err, ok := s.closeError.Load().(error); ok {
		return err
	}

	return nil
}

func (s *stream) accept(ctx context.Context, m *message) (err error) {
	id := m.Report.FeedID.String()

	s.waterMarkMu.Lock()
	if s.waterMark[id] >= m.Report.ObservationsTimestamp {
		s.stats.skipped.Add(1)
		s.waterMarkMu.Unlock()
		return nil
	}

	s.stats.accepted.Add(1)
	s.waterMark[id] = m.Report.ObservationsTimestamp
	s.waterMarkMu.Unlock()

	select {
	case <-ctx.Done():
		return ctx.Err()
	case s.output <- m.Report:
		return nil
	}
}

type wsConn struct {
	mu     sync.Mutex
	host   string
	origin string
	conn   *websocket.Conn
}

func (ws *wsConn) close() (err error) {
	ws.mu.Lock()
	defer ws.mu.Unlock()
	return ws.conn.CloseNow()
}

func (ws *wsConn) read(ctx context.Context, closingMutex *sync.RWMutex, accept func(context.Context, *message) error) (err error) {
	var lastErr error
	for {
		// coordinates with a potential Close function call from client
		closingMutex.RLock()
		_, b, err := ws.conn.Read(ctx)
		if err != nil {
			lastErr = err
			break
		}

		m := &message{}
		if err = json.Unmarshal(b, m); err != nil {
			lastErr = err
			break
		}

		if err = accept(ctx, m); err != nil {
			lastErr = err
			break
		}
		closingMutex.RUnlock()
	}
	closingMutex.RUnlock()
	return lastErr
}

func (ws *wsConn) replace(c *websocket.Conn) {
	ws.mu.Lock()
	defer ws.mu.Unlock()
	ws.conn = c
}

func (s *stream) newWSconn(ctx context.Context, origin string) (ws *wsConn, err error) {
	reqURL := s.config.wsURL.ResolveReference(&url.URL{Path: apiV1WS})
	reqURL.RawQuery = url.Values{"feedIDs": {strings.Join(feedIdsToStringList(s.feedIDs), ",")}}.Encode()

	headers := http.Header{}
	generateAuthHeaders(headers, http.MethodGet, reqURL.RequestURI(), nil,
		s.config.ApiKey, s.config.ApiSecret, time.Now().UnixMilli())

	if origin != "" {
		headers.Add(cllOriginHeader, origin)
	}

	if len(s.customHeaders) > 0 {
		for k, v := range s.customHeaders {
			headers.Add(k, v[0])
		}
	}

	opts := &websocket.DialOptions{
		HTTPHeader:      headers,
		CompressionMode: websocket.CompressionContextTakeover,
		HTTPClient:      s.httpClient,
		Host:            s.customHeaders.Get("Host"),
	}
	s.config.logDebug("client: stream websocket dial request url: %s, opts: %s", reqURL.String(), opts)
	conn, resp, err := websocket.Dial(ctx, reqURL.String(), opts)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode >= http.StatusBadRequest {
		return nil, fmt.Errorf("client: invalid status code %d", resp.StatusCode)
	}

	ws = &wsConn{
		host:   reqURL.Host,
		origin: origin,
		conn:   conn,
	}

	return ws, nil
}
